File: qc-core/src/main/java/com/eager/questioncloud/core/domain/payment/coupon/implement/UserCouponRegister.java
package com.eager.questioncloud.core.domain.payment.coupon.implement;

import com.eager.questioncloud.core.domain.payment.coupon.model.Coupon;
import com.eager.questioncloud.core.domain.payment.coupon.model.UserCoupon;
import com.eager.questioncloud.core.domain.payment.coupon.repository.UserCouponRepository;
import com.eager.questioncloud.core.exception.CustomException;
import com.eager.questioncloud.core.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
public class UserCouponRegister {
    private final CouponReader couponReader;
    private final CouponUpdater couponUpdater;
    private final UserCouponRepository userCouponRepository;

    @Transactional
    public UserCoupon registerCoupon(Long userId, String couponCode) {
        Coupon coupon = couponReader.findByCode(couponCode);

        if (userCouponRepository.isRegistered(userId, coupon.getId())) {
            throw new CustomException(Error.ALREADY_REGISTER_COUPON);
        }

        couponUpdater.decreaseCount(coupon.getId());

        return userCouponRepository.save(UserCoupon.create(userId, coupon));
    }
}

File: qc-core/qc-pay/qc-pay-api/src/main/kotlin/com/eager/questioncloud/payment/coupon/implement/UserCouponRegister.kt
package com.eager.questioncloud.payment.coupon.implement

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.payment.domain.UserCoupon
import com.eager.questioncloud.payment.repository.CouponInformationRepository
import com.eager.questioncloud.payment.repository.UserCouponRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class UserCouponRegister(
    private val couponInformationRepository: CouponInformationRepository,
    private val userCouponRepository: UserCouponRepository,
) {
    @Transactional
    fun registerCoupon(userId: Long, couponCode: String) {
        val coupon = couponInformationRepository.findByCode(couponCode)
        
        if (userCouponRepository.isRegistered(userId, coupon.id)) {
            throw CoreException(Error.ALREADY_REGISTER_COUPON)
        }
        
        if (!couponInformationRepository.decreaseCount(coupon.id)) {
            throw CoreException(Error.LIMITED_COUPON)
        }
        
        userCouponRepository.save(UserCoupon.create(userId, coupon))
    }
}
